finetune/adapter.py

```diff
@@ -117,13 +140,22 @@ def train(
 
         input_ids, targets = get_batch(fabric, train_data)
 
-        with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_steps != 0)):
+        if fabric.device.type == "xla":
             logits = model(input_ids)
             loss = loss_fn(logits, targets)
             fabric.backward(loss / gradient_accumulation_steps)
+        else:
+            with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_steps != 0)):
+                logits = model(input_ids)
+                loss = loss_fn(logits, targets)
+                fabric.backward(loss / gradient_accumulation_steps)
 
         if (iter_num + 1) % gradient_accumulation_steps == 0:
-            optimizer.step()
+            if fabric.device.type == "xla":
+                xm.optimizer_step(optimizer)
```

Comment:
Same question. Why can't we `optimizer.step()`?

Reply:
[Here](https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L878-L911) is the src code for `xm.optimizer_step()`.

Using `xm.optimizer_step()` will also reduce the gradient. Otherwise we need to include `xm.reduce_gradients(optimizer)`

Reply:
In Fabric, `optimizer.step()` is wrapped by the `_FabricOptimizer`. This wrappers routes the call to `Strategy.optimizer_step` which itself routes it to `Precision.optimizer_step`.

Particularly, XLA on Fabric uses the `XLAPrecision` class which does exactly this:
https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/plugins/precision/xla.py

Reply:
Hm, are we sure that the `xla` strategy wraps the optimizer? From what I can tell it is not wrapped by anything. I can create a PR in lightning src code if need be.

Reply:
For reference when setting up everything, `XLAStrategy` and `ParallelStrategy` do not have any functions for setting up the optimizers so we default to `setup_optimizer()` in the base `Strategy` class: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/strategies/strategy.py#L149.

Reply:
This happens through `fabric.setup`: https://github.com/Lightning-AI/lightning/blob/master/src/lightning/fabric/fabric.py#L223

If you are positive that we are not doing this wrapper correctly and calling `xm.optimizer_step()`, then there's a bug to fix upstream

Reply:
Okay, the issue was that the precision for xla must be set to "32-true", but by default lit-parrot defines precision as "bf16-true". With "32-true" `xm.optimizer_step()` will be called as expected. I will make the excepted change.